// rtt-engine.io/server.js
"use strict";

/*!
 * Engine.io - Realtime Transport Client
 * Copyright(c) 2013 Owen Barnes <owen@socketstream.org>
 * MIT Licensed
 */

var engine = require('engine.io');

module.exports = function(options) {

  return function(server) {

    var io = engine.listen(server.port, options);

    io.on('connection', function(socket) {
      
      // Grab it here as socket.requests gets removed when transport is upgraded
      var remoteAddress = socket.request.connection.remoteAddress;
  
      server.events.emit('connect', socket.id);

      // Process incoming messages
      socket.on('message', function(message){

        var request = {
          message:    message,
          transport:  'engineio',
          socketId:   socket.id,
          clientIp:   remoteAddress
        };

        server.process(request);

      });

      // Notify RTT when a client disconnects
      socket.on('close', function() {
        server.events.emit('disconnect', socket.id);
      });

    });

    return {

      sendToSocketId: function(socketId, msg) {
        io.clients[socketId].send(msg);
      },

      broadcast: function(msg){
        for (var socketId in io.clients) {
          io.clients[socketId].send(msg);
        }
      }
    };

  };
};
